### ISSUE TITLE
Need to be able to supply values for variables.

### ISSUE THREAD

**Issue by lrbarber** [apache/daffodil-vscode#200]
When running Daffodil from the command line, it is possible to supply values for variables using:
```
-D VARIABLE=VALUE 
```
A method to provide these values when working with the debugger is needed.

**Comment by lrbarber** [comment]
```
-T TUNABLE=VALUE
```

**Comment by mbeckerle** [comment]
To clarify, we do need to be able to specify tunables and variables in the UI, without question. 

But replicating the options the CLI can specify so that you also have those in the GUI is not enough. 

This feature should not be about parameterizing a run only, (as it is in the CLI) but about authoring, with GUI support, the config file that captures those parameters for repeated use from either the GUI or the command line, or in an application program that uses the Daffodil API. 

(Similarly to testing a schema should serve double-duty both as a way of doing the test, and of authoring the TDML file so the test is captured and repeatable. There's a separate issue/ticket for this.)

Regardless of all of the above, the important point is that the place all of Daffodil gets the list of say, tunables, and the list of all the suppressable warnings, is by processing the XSD for config files. The scala code gets generated from that by the daffodil-propgen module. 

Daffodil propgen can also simultaneously create whatever typescript, or json data or whatever the GUI needs to be able to support a nice UI for properties and warning suppression etc. All the per-tunable and per warning-flag boilerplate can be generated. 

We add new tunables and new warning flags all the time, so generating this code for the UI automatically would help keep things in sync between the engine and the UI.

**Comment by lrbarber** [comment]
That clarification helps, but we do still want to the ability to parameterize runs.

**Comment by mbeckerle** [comment]
Also see: https://issues.apache.org/jira/browse/DAFFODIL-2705 which is a JIRA ticket for daffodil. 

Turns out there is no single API call to just hand Daffodil a config file. Grrrrr. (Wishful thinking on my part.)

 You can hand it a config file for external variable bindings, but there is no API call for it to take tunables from a config file. Tunables can only be specified via API call taking a Map data structure, or as individual tunable settings one by one.

**Comment by arosien** [comment]
@lrbarber @mbeckerle is it most useful to define the variables/tunables before the run (supplied by config file, etc.), or are you preferring something definable at runtime (which may be tedious to define one-by-one)? I'm guessing the former, but just trying to clarify.

**Comment by lrbarber** [comment]
I agree that it could become tiresome having to reenter values each time, so a config file is probably the best option.

**Comment by arosien** [comment]
Other comments aside, I'm a bit unsure of the configuration semantics of variables. In my bogus example above, the schema doesn't have a `foo` variable, and daffodil rejects the parse because `foo` isn't defined in the schema, or wherever variables can be defined. Is that the intended validation semantics?

In other programs, undefined configuration parameters are ignored rather than rejected with an error.

**Comment by lrbarber** [comment]
Here's an example variable definition. Note that without the external="true", supplying a value from the command line (or config file, in this case) is invalid. I'm not sure what should be done in that case.


<dfdl:defineVariable name="test_str" type="xs:string" external="true" defaultValue="" />

**Comment by mbeckerle** [comment]
I would prefer to give the location of a config file, and have a standard XML editor for config files. They are just XML, and there is an XML Schema for what they can contain. 

This file is part of the schema, not part of the configuration of a UI. If you deploy the schema, this config file is part of what is packaged with the schema. 

If you replicate the config file contents into the UI launch config, then you need a way to save them out to a file when the schema is being saved, etc.

**Comment by mbeckerle** [comment]
When you try to set a variable from the API, the schema rejects an attempt to set a variable that doesn't exist. I believe it will also fail if the value (a string) you supply is not convertible to the type of the variable. 

Often these variables are quite important and critical things like whether certain situations are to be deemed errors in the data or not.
